src/sleep.rs

```diff
@@ -0,0 +1,763 @@
+            RtcWakeLevel::AnyHigh => esp_sleep_ext1_wakeup_mode_t_ESP_EXT1_WAKEUP_ANY_HIGH,
+        }
+    }
+}
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub trait RtcPinDriver {}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<'d, P, M> RtcPinDriver for PinDriver<'d, P, M>
+where
+    P: RTCPin + InputPin + 'd,
+    M: RTCMode + 'd,
+{
+}
+
+pub trait RtcWakeupPinTrait {
+    type PinDriver;
+    fn pin(&self) -> i32;
+}
+
+pub struct DummyRtcWakeupPin;
+impl RtcWakeupPinTrait for DummyRtcWakeupPin {
+    type PinDriver = DummyRtcWakeupPin;
+    fn pin(&self) -> i32 {
+        0
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub trait RtcWakeupPins {
+    type WakeupPin: RtcWakeupPinTrait;
+    type Iterator<'a>: Iterator<Item = i32>
+    where
+        Self: 'a;
+
+    fn iter(&self) -> Self::Iterator<'_>;
+}
+#[cfg(not(any(esp32, esp32s2, esp32s3)))]
+pub trait RtcWakeupPins {}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<P> RtcWakeupPins for P
+where
+    P: RtcWakeupPinTrait,
+    P::PinDriver: RtcPinDriver,
+{
+    type WakeupPin = P;
+
+    type Iterator<'a> = core::iter::Once<i32> where Self: 'a;
+
+    fn iter(&self) -> Self::Iterator<'_> {
+        core::iter::once(self.pin())
+    }
+}
+
+pub struct EmptyRtcWakeupPins<A>(PhantomData<A>);
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<A> EmptyRtcWakeupPins<A> {
+    pub fn chain<O>(other: O) -> ChainedRtcWakeupPins<Self, O>
+    where
+        A: RtcWakeupPinTrait,
+        O: RtcWakeupPins<WakeupPin = A>,
+    {
+        ChainedRtcWakeupPins {
+            first: Self(PhantomData),
+            second: other,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<A> RtcWakeupPins for EmptyRtcWakeupPins<A>
+where
+    A: RtcWakeupPinTrait,
+{
+    type WakeupPin = A;
+
+    type Iterator<'a> = core::iter::Empty<i32> where Self: 'a;
+
+    fn iter(&self) -> Self::Iterator<'_> {
+        core::iter::empty()
+    }
+}
+
+#[cfg(not(any(esp32, esp32s2, esp32s3)))]
+impl<A> RtcWakeupPins for EmptyRtcWakeupPins<A> {}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+pub struct ChainedRtcWakeupPins<F, S> {
+    first: F,
+    second: S,
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<F, S> ChainedRtcWakeupPins<F, S> {
+    pub fn chain<O>(self, other: O) -> ChainedRtcWakeupPins<Self, O>
+    where
+        F: RtcWakeupPins,
+        S: RtcWakeupPins,
+        O: RtcWakeupPins,
+    {
+        ChainedRtcWakeupPins {
+            first: self,
+            second: other,
+        }
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+impl<F, S> RtcWakeupPins for ChainedRtcWakeupPins<F, S>
+where
+    F: RtcWakeupPins,
+    S: RtcWakeupPins,
+{
+    type WakeupPin = F::WakeupPin;
+
+    type Iterator<'a> = core::iter::Chain<F::Iterator<'a>, S::Iterator<'a>> where Self: 'a;
+
+    fn iter(&self) -> Self::Iterator<'_> {
+        self.first.iter().chain(self.second.iter())
+    }
+}
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
```

Comment:
If we can get rid of the `RtcWakeupPinTrait` I think we can get rid of this struct as well.